bluesky/preprocessors.py

```diff
@@ -1307,20 +1307,25 @@ def __call__(self, plan):
         return (yield from plan)
 
 
-def define_run_wrapper(plan, run):
+def set_run_name_wrapper(plan, run):
     """
     Add a run id to each message in wrapped plan
 
     Parameters
     ----------
     plan : iterable or iterator
         a generator, list, or similar containing `Msg` objects
-    run_id : str
-        The run to set on each Msg
+    run : str
+        The run name to set on each Msg
     """
+
+    if not isinstance(run, str):
+        raise ValueError(f"run name must be a string: passed value {run} is of type {type(run)}")
     def _set_run_id(msg):
-        return msg._replace(run=run)
+        if not isinstance(msg.run, str):
```

Comment:
I would prefer that this be `if msg.run is None`.  It is _marginally_ more performant and means things will still work if the user provides some none-string input.  I think that, for things to work, we only really need this to be hashable.

The `object()` singleton is only droped in once the RE gets ahold of this.  On a bit more reflection, I think the logic in `_extract_run_key` is not needed.  I think it is a hold over from an earlier theory when we considered putting the run id in the kwargs.

Reply:
Yes, I think it can be done this way as well. In this case ``_extract_run_key`` should simply return ``msg.run``, or be completely removed. Then user provided run name should be only rejected if it is ``None``. Another question if the right name should be 'run name' or 'run ID'. The latter may be easily mixed with scan ID.

Reply:
Removing ``_extract_run_key`` completely may make it difficult to change the behavior in the future.

Reply:
I assume that if ``run_id`` may have any type but ``None``, the wrapper/decorator should be renamed into ``set_run_id``. The user documentation should recommend to use strings in manually created plans for better readability.